Fetch groups in a Google Connector

JoelSpeed opened the pull request:
Fixes #866 #1001 

**This PR should not be used without** #1180. Wanted to get the PR opened though in case anyone wants this feature urgently and can't wait until both PRs are merged. 

I have created a separate Google connector that uses a service account to fetch all groups that a user is a member of. 

I wasn't sure which documentation would need to be updated so I have added a document explaining the new parameters and how to set up the Google service account, but I reckon this will need extra documentation changes before it can be merged, let me know what it needs.

jhohertz commented:
Just wanted to note I am presently testing this variation (plus #1180) and find it to be a valuable addition.

What needs to happen to have the code mainlined? Happy to help in any way I can.

pdh commented on connector/google/google.go:
```diff
@@ -0,0 +1,266 @@
+
+	var userGroups []string
+	for _, group := range groupsList.Groups {
+		userGroups = append(userGroups, group.Email)
```
For this func to return the groups the user is a member of should this be the group's name instead of email? ie. `group.Name`

JoelSpeed replied:
I chose to use the group's email field since this is globally unique. 

I considered the groups name's have to be unique within Google but they may clash with other groups defined within Kubernetes via other authenticators (intentionally or not) so it seemed safer to use the email address

ericchiang replied:
Maybe this should be configurable? (a TODO is fine)

JoelSpeed replied:
Added a TODO for now

ericchiang commented on connector/google/google.go:
```diff
@@ -0,0 +1,266 @@
+// Config holds configuration options for Google logins.
+type Config struct {
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
```
nit: can this be optionally read from a file? e.g. `ClientSecretFile`

JoelSpeed replied:
I don't see this being done in any of the other connectors but could add this. I assume the `ClientSecretFile` would just contain a single string? Or is there a format for files from Google containing the `ClientSecret`?

ericchiang commented on connector/google/google.go:
```diff
@@ -0,0 +1,266 @@
+// getGroups creates a connection to the admin directory service and lists
+// all groups the user is a member of
+func (c *googleConnector) getGroups(email string) ([]string, error) {
+	srv, err := createDirectoryService(c.serviceAccountFilePath, c.adminEmail)
```
Why is this done on every getGroups call instead of on init?

JoelSpeed replied:
That's a very good point, will fix

JoelSpeed commented:
I've rebased and addressed a couple of your comments @ericchiang 

Took me a while to work out `go mod` as I haven't used it before but I think I got it. I need to squash the commits down before we merge this, should I prefix the commit messages based on where the commit touches?

tsuna commented:
ping pretty please :)

ericchiang commented:
Still blocked on https://github.com/dexidp/dex/pull/1180, right?

JoelSpeed commented:
Still blocked on #1180 indeed, apologies all, will get back on to that once I'm recovered from KubeCon 😴

surajssd commented:
It would be awesome to have this feature in dex master! Really waiting for this to be merged!

jhohertz commented:
I'd like to test this w/ #1180 , any way we could see this rebased?

JoelSpeed commented:
@jhohertz I've just rebased this branch, you should be able to test with 1180 with a little work

jhohertz commented:
Testing the rebased PR... This is working well along with #1180 

What do we need to do to see this get merged?

quantonganh commented:
#1180 has been merged.

@ericchiang  @bonifaido Do you have some time to take a review?

bonifaido commented:
Just checking this.

justaugustus commented:
@srenatus @rithujohn191 -- Can you able a look at this one?
